src/stores/room-list/filters/VisibilityProvider.ts

```diff
@@ -38,7 +39,7 @@ export class VisibilityProvider {
         await VoipUserMapper.sharedInstance().onNewInvitedRoom(room);
     }
 
-    public isRoomVisible(room?: Room): boolean {
+    public async isRoomVisible(room?: Room): Promise<boolean> {
```

Comment:
This is a breaking change, existing customisations will fail to compile with this change due to the return type changing.

Reply:
What would you suggest?

Reply:
Move it into the js-sdk as a background operation, so consumers don't have to make all their APIs async. Feels a bit strange for `getRuleForInvite` to be an async function which actually creates rooms, it seems like that one should be read-only, and something else should create and uphold the policy rooms, emitting "PolicyUpdated" or similar events for consumers to subscribe and react to.

Reply:
It also looks like `getOrCreateTargetRoom` is flawed if called multiple times whilst the room is being created, it'll stack multiple creations instead of being idempotent. This could also occur with this PR where multiple operations overlap and call isRoomVisible causing overlapping calls to `getOrCreateTargetRoom`

Reply:
> It also looks like `getOrCreateTargetRoom` is flawed if called multiple times whilst the room is being created, it'll stack multiple creations instead of being idempotent. This could also occur with this PR where multiple operations overlap and call isRoomVisible causing overlapping calls to `getOrCreateTargetRoom`

You're right, I should at least lock it to avoid that.

Reply:
> Move it into the js-sdk as a background operation, so consumers don't have to make all their APIs async. Feels a bit strange for `getRuleForInvite` to be an async function which actually creates rooms, it seems like that one should be read-only, and something else should create and uphold the policy rooms, emitting "PolicyUpdated" or similar events for consumers to subscribe and react to.

That was the design at some point but I dropped it because it felt like I would be reimplementing a non-trivial subset of the state event store. On the other hand, it is probably needed to achieve any kind of performance anyway, so now might be the right time to revisit this.

I'll think this over.

Reply:
In case someone other reads this, this conversation has spawned a patch in matrix-js-sdk: https://github.com/matrix-org/matrix-js-sdk/pull/2653 .